text/0061-functions.md

```diff
@@ -0,0 +1,295 @@
+Moreover, custom errors from constructor parameter validity checks can be emitted during validation when using extension functions, but not when using Cedar functions.
+
+Second, there is no special parsing for Cedar functions. With Cedar's built-in `decimal`, you can write `decimal("123.12")` which more directly conveys the number being represented than does `mydecimal(123,12)`.
+
+Of course, these drawbacks do not necessarily speak against Cedar functions generally, but suggest that for suitably general use-cases (like decimal numbers!), an extension function might be warranted instead.
+
+### Readability: Policies are no longer standalone
+A policy can no longer be read by itself, it has to be read in the context of all function definitions it uses.
+Policies that use a large number of functions may be hard to read.
+
+## Alternatives
+
+### Type annotations
+We could require/allow Cedar functions to have type annotations, taking type definitions from either the schema or allowing them to be inline.
+
+Example:
+
+Schema:
+```
+type SemVer = { major : Long, minor : Long, patch : Long };
+```
+Policy:
+```
+function semver(major : Long, minor : Long, patch : Long) -> Semver {
+  { major : major, minor : minor, patch : patch }
+};
+```
+
+This would allow functions to typechecked in absence of a use-site, and allow for easier user specification of intent.
+It would also probably result in clearer type checker error messages.
+
+This introduces the following questions:
+
+1. Do we allow `type` declaration in policies, or just in schemas?
+2. Are type annotations enforced dynamically à la Contracts, or are they just ignored at runtime?
+  1. If they are dynamically enforced, that implies access to the schema to unfold type definitions. It also may introduce redundant type checking.
+3. Are type annotations required or optional?
+4. Will we have types to support generics, i.e., polymorphism?
+
+
+Leaving them out for this RFC only precludes only one future design decision, making type annotations required.
+This decision feels unlikely, as we want Cedar to be useful without using a schema/the validator.
+Adding _optional_ type annotations in the future is backwards compatible, but mandating type annotations may not be, due to the use of polymorphic functions.
+
+### Let functions call other functions
+As long as cycles are forbidden and functions as arguments are disallowed, we could allow functions to call other functions without sacrificing termination.
+However, the potential complexity explosion is high, and it's backwards compatible to add this later.
+
+<!-- ### Make parse errors runtime errors
+Make (any subset of) the following errors runtime errors instead of parse errors: 
+
+1. Application of non-function
+2. Non-application of function
+3. Function application with incorrect arity -->
```

Comment:
I commented this out because it was hard for me to imagine why you would want this. You can always detect these problems perfectly (no false alarms) at compile-time, right?

Reply:
Unless we add first class functions, or some kind of late-binding mechanism

Reply:
What do you mean by "late binding" mechanism? In any case, we have neither of these so no need to discuss until we are actually proposing to have them (we can always relax the parser later, if we add them).